File: packages/igloo-ts/src/widgets/grid.ts
import { Grid as WitGrid } from 'iced:app/grid@0.1.0';
import type { Pixels } from 'iced:app/shared@0.1.0';
import { Element, toElement, type ElementLike, type IntoElement } from '../element.js';

/**
 * Builder for creating Grid layout widgets.
 * A Grid arranges its contents in a grid pattern.
 *
 * @example
 * ```typescript
 * const grid = Grid.new()
 *   .columns(3)
 *   .spacing(10)
 *   .push(Text.new('Cell 1'))
 *   .push(Text.new('Cell 2'));
 * ```
 *
 * The child message type is inferred and accumulated as elements are pushed;
 * pass `Grid.new<Msg>()` to pin it so each `push` is checked against `Msg`.
 *
 * @typeParam Msg - The accumulated message type of the grid's children.
 * @typeParam Bound - Upper bound each child must satisfy; `unknown` when unpinned.
 */
export class Grid<Msg = never, Bound = unknown> implements IntoElement<Msg> {
  private raw: WitGrid;

  private constructor() {
    this.raw = new WitGrid();
  }

  /** Create a new empty Grid builder */
  static new(): Grid<never, unknown>;
  /** Create a new empty Grid builder pinned to the message type `B` */
  static new<B>(): Grid<never, B>;
  static new(): Grid<never, unknown> {
    return new Grid();
  }

  /** Create a Grid with the given elements */
  static with<const M>(elements: ElementLike<M>[]): Grid<M, unknown> {
    return new Grid<M, unknown>().extend(elements);
  }

  /** Add an element to the grid */
  push<const M extends Bound>(element: ElementLike<M>): Grid<Msg | M, Bound> {
    this.raw.push(toElement(element).inner);
    return this as unknown as Grid<Msg | M, Bound>;
  }

  /** Add multiple elements */
  extend<const M extends Bound>(elements: ElementLike<M>[]): Grid<Msg | M, Bound> {
    for (const element of elements) {
      this.raw.push(toElement(element).inner);
    }
    return this as unknown as Grid<Msg | M, Bound>;
  }

  /** Set the spacing between cells */
  spacing(spacing: Pixels): this {
    this.raw.spacing(spacing);
    return this;
  }

  /** Set the grid width in pixels */
  width(width: Pixels): this {
    this.raw.width(width);
    return this;
  }

  /** Set the grid height in pixels */
  height(height: Pixels): this {
    this.raw.height(height);
    return this;
  }

  /** Set the number of columns */
  columns(columns: number | bigint): this {
    this.raw.columns(BigInt(columns));
    return this;
  }

  /** Set fluid column width (columns resize based on content) */
  fluid(width: Pixels): this {
    this.raw.fluid(width);
    return this;
  }

  /** Convert to Element */
  intoElement(): Element<Msg> {
    return new Element(WitGrid.intoElement(this.raw));
  }
}

File: crates/igloo_guest/src/widgets/grid.rs
use std::marker::PhantomData;

use iced_core::Pixels;

use crate::Element;
use crate::bindings::iced::app::widgets::{GridNode, Node};

/// A container that arranges its contents in a grid.
pub struct Grid<Message> {
    children: Vec<Element<Message>>,
    spacing: Option<f32>,
    width: Option<f32>,
    height: Option<f32>,
    columns: Option<u64>,
    fluid: Option<f32>,
    _message: PhantomData<Message>,
}

impl<Message> Default for Grid<Message> {
    fn default() -> Self {
        Self::new()
    }
}

impl<Message> Grid<Message> {
    /// Creates an empty [`Grid`].
    pub fn new() -> Self {
        Self {
            children: Vec::new(),
            spacing: None,
            width: None,
            height: None,
            columns: None,
            fluid: None,
            _message: PhantomData,
        }
    }

    /// Creates a [`Grid`] with the given capacity.
    pub fn with_capacity(_capacity: usize) -> Self {
        Self::new()
    }

    /// Creates a [`Grid`] with the given elements.
    pub fn with_children(children: impl IntoIterator<Item = Element<Message>>) -> Self {
        Self::new().extend(children)
    }

    /// Creates a [`Grid`] from an already allocated [`Vec`].
    pub fn from_vec(children: Vec<Element<Message>>) -> Self {
        Self::new().extend(children)
    }

    /// Adds an element to the [`Grid`].
    pub fn push(mut self, child: impl Into<Element<Message>>) -> Self {
        self.children.push(child.into());
        self
    }

    /// Extends the [`Grid`] with the given children.
    pub fn extend(self, children: impl IntoIterator<Item = Element<Message>>) -> Self {
        children.into_iter().fold(self, Self::push)
    }

    /// Sets the spacing between cells in the [`Grid`].
    pub fn spacing(mut self, amount: impl Into<Pixels>) -> Self {
        self.spacing = Some(amount.into().0);
        self
    }

    /// Sets the width of the [`Grid`].
    pub fn width(mut self, width: impl Into<Pixels>) -> Self {
        self.width = Some(width.into().0);
        self
    }

    /// Sets the height of the [`Grid`].
    pub fn height(mut self, height: impl Into<Pixels>) -> Self {
        self.height = Some(height.into().0);
        self
    }

    /// Sets the number of columns of the [`Grid`].
    pub fn columns(mut self, columns: u64) -> Self {
        self.columns = Some(columns);
        self
    }

    /// Sets the fluid spacing of the [`Grid`].
    pub fn fluid(mut self, amount: impl Into<Pixels>) -> Self {
        self.fluid = Some(amount.into().0);
        self
    }
}

impl<Message: 'static> From<Grid<Message>> for Element<Message> {
    fn from(grid: Grid<Message>) -> Self {
        Element::new(move |realize, arena| {
            let elements = grid
                .children
                .into_iter()
                .map(|child| child.build(realize, arena))
                .collect();
            let node = GridNode {
                elements,
                spacing: grid.spacing,
                width: grid.width,
                height: grid.height,
                columns: grid.columns,
                fluid: grid.fluid,
            };
            arena.push(Node::Grid(node))
        })
    }
}
